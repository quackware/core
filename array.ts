/**
 * Create a new object with an index created by the provided {@link indexer}
 */
export function index<T>(
  array: ReadonlyArray<T>,
  indexer: (t: T) => string,
): { [key: string]: T };
export function index<T, R>(
  array: ReadonlyArray<T>,
  indexer: (t: T) => string,
  mapper: (t: T) => R,
): { [key: string]: R };
export function index<T, R>(
  array: ReadonlyArray<T>,
  indexer: (t: T) => string,
  mapper?: (t: T) => R,
): { [key: string]: R } {
  return array.reduce((r, t) => {
    r[indexer(t)] = mapper ? mapper(t) : t;
    return r;
  }, Object.create(null));
}

export { distinctBy } from "https://deno.land/std@0.210.0/collections/distinct_by.ts";
