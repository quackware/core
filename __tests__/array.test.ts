import { index } from "../array.ts";
import { assertSnapshot } from "./test-deps.ts";

Deno.test("array", async (t) => {
  const data = ["one", "two", "three"] as const;

  await t.step("index", async (t) => {
    const indexed = index(data, function idx(t) {
      return t.toString();
    }, function mapper(t) {
      return { type: t };
    });

    await assertSnapshot(t, indexed);
  });
});
