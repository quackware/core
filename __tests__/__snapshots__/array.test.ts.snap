export const snapshot = {};

snapshot[`array > index 1`] = `
{
  one: {
    type: "one",
  },
  three: {
    type: "three",
  },
  two: {
    type: "two",
  },
}
`;
